cmd/coral.go

```diff
@@ -0,0 +1,383 @@
+provisioning stages.`,
+	RunE: func(cmd *coral.Command, args []string) error {
+		validateCmdConfigFlag, _ := cmd.Flags().GetString("config")
+		validateCmdAdapterFlag, _ := cmd.Flags().GetString("adapter")
+
+		input, _, err := loadConfig(validateCmdConfigFlag, validateCmdAdapterFlag)
+		if err != nil {
+			return err
+		}
+		input = caddy.RemoveMetaFields(input)
+
+		var cfg *caddy.Config
+		err = json.Unmarshal(input, &cfg)
+		if err != nil {
+			return fmt.Errorf("decoding config: %v", err)
+		}
+
+		err = caddy.Validate(cfg)
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("Valid configuration")
+
+		return nil
+	},
+}
+
+var fmtCmd = &coral.Command{
+	Use:   "fmt [--overwrite] [config]",
+	Args:  coral.MaximumNArgs(1),
+	Short: "Formats a Caddyfile",
+	Long: `
+Formats the Caddyfile by adding proper indentation and spaces to improve
+human readability. It prints the result to stdout.
+
+If --overwrite is specified, the output will be written to the config file
+directly instead of printing it.
+
+If you wish you use stdin instead of a regular file, use - as the path.
+When reading from stdin, the --overwrite flag has no effect: the result
+is always printed to stdout.`,
+	RunE: func(cmd *coral.Command, args []string) error {
+		formatCmdConfigFile := ""
+		if len(args) > 0 {
+			formatCmdConfigFile = args[1]
+		}
+		if formatCmdConfigFile == "" {
+			formatCmdConfigFile = "Caddyfile"
+		}
+
+		overwrite, _ := cmd.Flags().GetBool("overwrite")
+
+		// as a special case, read from stdin if the file name is "-"
+		if formatCmdConfigFile == "-" {
+			input, err := io.ReadAll(os.Stdin)
+			if err != nil {
+				return fmt.Errorf("reading stdin: %v", err)
+			}
+			fmt.Print(string(caddyfile.Format(input)))
+			return nil
+		}
+
+		input, err := os.ReadFile(formatCmdConfigFile)
+		if err != nil {
+			return fmt.Errorf("reading input file: %v", err)
+		}
+
+		output := caddyfile.Format(input)
+
+		if overwrite {
+			if err := os.WriteFile(formatCmdConfigFile, output, 0600); err != nil {
+				return fmt.Errorf("overwriting formatted file: %v", err)
+			}
+		} else {
+			fmt.Print(string(output))
+		}
+
+		return nil
+	},
+}
+
+var runCmd = &coral.Command{
+	Use: "run",
+}
+var startCmd = &coral.Command{}
+var reloadCmd = &coral.Command{}
+var stopCmd = &coral.Command{}
+
+var upgradeCmd = &coral.Command{}
+var addPackageCmd = &coral.Command{}
+var removePackageCmd = &coral.Command{}
+
+// TODO: figure out the correct directory
+var manpageCmd = &coral.Command{
+	Use:   "manpage",
+	Short: "Generate the man pages",
+	Run: func(_ *coral.Command, _ []string) {
+		doc.GenManTree(rootCmd, &doc.GenManHeader{
+			Title:   "Caddy",
+			Section: "1",
+		}, "./")
+	},
+}
+
+func init() {
+	listModulesCmd.Flags().Bool("packages", false, "Print package paths")
+	listModulesCmd.Flags().Bool("versions", false, "Print version information")
+	listModulesCmd.Flags().Bool("skip-standard", false, "Skip printing standard modules")
```

Comment:
Hmm, this is an awkward API. It doesn't have a way to define the flags along with the command `coral.Command` higher up? feels a bit detached to have to define these separately.

Reply:
Yeah, I didn't like the ergonomics of it much. It doesn't seem to be possible from the docs, for one `flags` is a private field of `coral.Command`, but it might be cleaner if I figure out how to adapt `caddycmd.Command` to `*coral.Command` since the former has its flags exposed and available for reading.

Reply:
And now?

Reply:
Nice! Yeah that looks much better.